src/main/c/Makefile

```diff
@@ -2,29 +2,84 @@
 
 # If you want to add a new cpp file, like foo.cpp, to the library, add foo.o to
 # this list
-OBJECTS=ibs.o davies.o
-
-CXX ?= c++
-# append existing flags so they override our flags
-CXXFLAGS += -O3 -march=native -g -std=c++11 -Ilibsimdpp-2.0-rc2 -Wall -Werror
-CXXFLAGS += -fPIC -ggdb
-LIBFLAGS += -fvisibility=hidden
+OBJECTS := \
+  ibs.o davies.o \
+  NativeModule.o NativePtr.o \
+  NativeStatus.o NativeLongFunc.o NativeCodeSuite.o
 
 UNAME_S := $(shell uname -s)
 UNAME_P := $(shell uname -p)
+
+# Compile with new version of clang if it is installed
+ifeq ($(UNAME_S),Linux)
+  ifndef CLANG_HOME
+    CLANG_HOME := $(shell /bin/ls -d /usr/l*/llvm* /usr/l*/llvm-5* /usr/l*/llvm-6* 2>/dev/null | tail -1)
+  endif
+  ifndef JAVA_HOME
+    JAVA_HOME := $(shell /bin/ls -d /usr/lib/jvm/java-8-*64 /usr/lib/jvm/java-9-*64 2>/dev/null | tail -1)
+  endif
+  JAVA_MD := linux
+else
+  CLANG_HOME := /usr/local/llvm-$(CLANG_VERSION)
+  ifndef JAVA_HOME
+    JAVA_JDK := $(shell /bin/ls -d /Library/Java/JavaVirtualMachines/jdk1.8*.jdk 2> /dev/null | tail -1)
+    ifneq ($(JAVA_JDK),)
+      JAVA_HOME := $(JAVA_JDK)/Contents/Home
+    endif
+  endif
+  JAVA_MD := darwin
+endif
+
+ifeq ($(CLANG_HOME),)
+  CLANG_HOME := CLANG_HOME_undefined
+endif
+ifeq ($(JAVA_HOME),)
+  JAVA_HOME := JAVA_HOME_undefined
+endif
+
+ifeq ($(shell [ -d $(CLANG_HOME) ] && echo found),found)
+  CXX := $(CLANG_HOME)/bin/clang
+  CXXSTD := -std=c++14
+  LIBDIRS := -L$(CLANG_HOME)/lib
+else
+  CXX ?= c++
+  CXXSTD := -std=c++11
+  LIBDIRS :=
+endif
+
+ifeq ($(UNAME_S),Darwin)
+  # CXX := c++
+endif
+
+# Eventually we would like -std=c++17, but the CI machines have
+# compilers which only go to -std=c++11
+
+# append existing flags so they override our flags
+CXXFLAGS += -O3 -march=native -g $(CXXSTD) -Ilibsimdpp-2.0-rc2 -Wall -Werror
 ifeq ($(UNAME_S),Linux)
-       LIBFLAGS += -rdynamic -shared
-       shared_library = lib/linux-x86-64/libhail.so
-       ifneq ($(filter %86,$(UNAME_P)),)
-               shared_library = lib/linux-x86/libhail.so
-       endif
+#CXXFLAGS += --no-strict-aliasing
+endif
+CXXFLAGS += -fPIC -ggdb -fno-strict-aliasing
+CXXFLAGS += -I../resources/include -I$(JAVA_HOME)/include -I$(JAVA_HOME)/include/$(JAVA_MD)
+LIBFLAGS += -fvisibility=default
+
+ifeq ($(UNAME_S),Linux)
+  LIBFLAGS += -rdynamic -shared
+  shared_library = lib/linux-x86-64/libhail.so
+  ifneq ($(filter %86,$(UNAME_P)),)
+    shared_library = lib/linux-x86/libhail.so
+  endif
 endif
 ifeq ($(UNAME_S),Darwin)
```

Comment:
I'm inclined, if possible, to have one if-block for each architecture. Would be able to pull that off in this setting?